File: src/cache.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::time::{Duration, SystemTime, Instant};
use serde::{Serialize, Deserialize};
use crate::error::{ProcessorError, Result};
use std::fs;
use std::sync::Arc;
use tokio::sync::RwLock;
use tokio::fs as tokio_fs;
use chrono;
use std::io::Read;

/// Represents a single entry in the file cache
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CacheEntry {
    /// The URL or identifier for this cache entry
    pub url: String,
    /// The file path where the cached content is stored
    pub path: PathBuf,
    /// When this entry was created or last updated
    pub timestamp: SystemTime,
    /// How long this entry should remain valid
    pub ttl: Duration,
}

/// A file-based cache for storing downloaded or generated content
#[derive(Debug)]
pub struct FileCache {
    cache_dir: PathBuf,
    default_ttl: Duration,
}

impl FileCache {
    /// Creates a new FileCache with the specified cache directory and default TTL
    pub fn new(cache_dir: PathBuf, default_ttl: Duration) -> Self {
        Self {
            cache_dir,
            default_ttl,
        }
    }
    
    /// Retrieves a cached entry by its key if it exists and is not expired
    pub async fn get(&self, key: &str) -> Result<Option<CacheEntry>> {
        let path = self.get_path(key);
        
        if !path.exists() {
            return Ok(None);
        }
        
        let content = tokio::fs::read_to_string(&path)
            .await
            .map_err(|e| ProcessorError::IO(e))?;
        let entry: CacheEntry = serde_json::from_str(&content)?;
        
        // Check if entry is expired
        let now = SystemTime::now();
        if now.duration_since(entry.timestamp).unwrap_or_default() > entry.ttl {
            // Entry is expired, remove it
            tokio::fs::remove_file(&path)
                .await
                .map_err(|e| ProcessorError::IO(e))?;
            return Ok(None);
        }
        
        Ok(Some(entry))
    }
    
    /// Stores a value in the cache with the specified key and TTL
    pub async fn set(&self, key: &str, value: &[u8], ttl: Option<Duration>) -> Result<PathBuf> {
        let path = self.get_path(key);
        
        // Ensure cache directory exists
        if let Some(parent) = path.parent() {
            tokio::fs::create_dir_all(parent)
                .await
                .map_err(|e| ProcessorError::IO(e))?;
        }
        
        // Write the value
        tokio::fs::write(&path, value)
            .await
            .map_err(|e| ProcessorError::IO(e))?;
        
        // Create and save the entry metadata
        let entry = CacheEntry {
            url: key.to_string(),
            path: path.clone(),
            timestamp: SystemTime::now(),
            ttl: ttl.unwrap_or(self.default_ttl),
        };
        
        let meta_path = self.get_meta_path(key);
        tokio::fs::write(&meta_path, serde_json::to_string(&entry)?)
            .await
            .map_err(|e| ProcessorError::IO(e))?;
        
        Ok(path)
    }
    
    fn get_path(&self, key: &str) -> PathBuf {
        // Create a safe filename from the key
        let safe_key = key.replace(['/', ':', '?', '&', '='], "_");
        self.cache_dir.join(safe_key)
    }
    
    fn get_meta_path(&self, key: &str) -> PathBuf {
        let mut path = self.get_path(key);
        path.set_extension("meta");
        path
    }
}

/// A generic in-memory cache for any serializable type
#[derive(Debug)]
pub struct Cache<T> {
    store: Arc<RwLock<HashMap<String, (T, Instant)>>>,
    ttl: Duration,
}

impl<T: Clone + Send + Sync + 'static> Cache<T> {
    /// Creates a new in-memory cache with the specified TTL
    pub fn new(ttl: Duration) -> Self {
        Self {
            store: Arc::new(RwLock::new(HashMap::new())),
            ttl,
        }
    }

    /// Retrieves a value from the cache by its key
    pub async fn get(&self, key: &str) -> Option<T> {
        let store = self.store.read().await;
        store.get(key)
            .filter(|(_, time)| time.elapsed() < self.ttl)
            .map(|(value, _)| value.clone())
    }

    /// Stores a value in the cache with the specified key
    pub async fn set(&self, key: &str, value: T) {
        let mut store = self.store.write().await;
        store.insert(key.to_string(), (value, Instant::now()));
    }

    /// Removes an entry from the cache by its key
    pub async fn remove(&self, key: &str) -> bool {
        let mut store = self.store.write().await;
        store.remove(key).is_some()
    }

    /// Clears all entries from the cache
    pub async fn clear(&self) {
        let mut store = self.store.write().await;
        store.clear();
    }

    /// Checks if the cache contains the specified key
    pub async fn contains_key(&self, key: &str) -> bool {
        let store = self.store.read().await;
        store.contains_key(key)
    }

    /// Returns the number of entries in the cache
    pub async fn len(&self) -> usize {
        let store = self.store.read().await;
        store.len()
    }

    /// Checks if the cache is empty
    pub async fn is_empty(&self) -> bool {
        let store = self.store.read().await;
        store.is_empty()
    }
    
    /// Removes expired entries from the cache and returns the count of removed entries
    pub async fn cleanup_expired(&self) -> usize {
        let mut store = self.store.write().await;
        let before_len = store.len();
        store.retain(|_, (_, time)| time.elapsed() < self.ttl);
        before_len - store.len()
    }
    
    /// Sets a new TTL for the cache
    pub fn set_ttl(&mut self, ttl: Duration) {
        self.ttl = ttl;
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct StoredCacheEntry {
    value: String,
    created_at: u64,
}

/// A specialized cache for storing string values with persistent storage
#[derive(Debug)]
pub struct StringCache {
    store: Arc<RwLock<HashMap<String, (String, Instant)>>>,
    ttl: Duration,
    cache_dir: PathBuf,
}

impl StringCache {
    /// Creates a new StringCache with the specified cache directory
    pub async fn new(cache_dir: &Path) -> Result<Self> {
        tokio::fs::create_dir_all(cache_dir)
            .await
            .map_err(|e| ProcessorError::IO(e))?;
        
        Ok(Self {
            store: Arc::new(RwLock::new(HashMap::new())),
            ttl: Duration::from_secs(3600), // Default 1 hour TTL
            cache_dir: cache_dir.to_path_buf(),
        })
    }

    /// Creates a new StringCache with the specified cache directory and TTL
    pub async fn with_ttl(cache_dir: &Path, ttl: Duration) -> Result<Self> {
        tokio::fs::create_dir_all(cache_dir)
            .await
            .map_err(|e| ProcessorError::IO(e))?;
        
        Ok(Self {
            store: Arc::new(RwLock::new(HashMap::new())),
            ttl,
            cache_dir: cache_dir.to_path_buf(),
        })
    }

    /// Sets a new TTL for the cache
    pub fn set_ttl(&mut self, ttl: Duration) {
        self.ttl = ttl;
    }

    /// Retrieves a value from the cache by its key
    pub async fn get(&self, key: &str) -> Result<Option<String>> {
        let path = self.get_path(key);
        
        if path.exists() {
            // Use tokio::fs instead of std::fs
            let content = tokio::fs::read_to_string(&path)
                .await
                .map_err(|e| ProcessorError::IO(e))?;
            Ok(Some(content))
        } else {
            Ok(None)
        }
    }

    /// Stores a value in the cache with the specified key
    pub async fn set_value(&self, key: &str, value: String) {
        let mut store = self.store.write().await;
        store.insert(key.to_string(), (value, Instant::now()));
    }
    
    /// Removes an entry from the cache by its key
    pub async fn remove(&self, key: &str) -> bool {
        let mut store = self.store.write().await;
        store.remove(key).is_some()
    }

    /// Clears all entries from the cache
    pub async fn clear(&self) {
        let mut store = self.store.write().await;
        store.clear();
    }

    /// Checks if the cache contains the specified key
    pub async fn contains_key(&self, key: &str) -> bool {
        let store = self.store.read().await;
        store.contains_key(key)
    }

    /// Returns the number of entries in the cache
    pub async fn len(&self) -> usize {
        let store = self.store.read().await;
        store.len()
    }

    /// Checks if the cache is empty
    pub async fn is_empty(&self) -> bool {
        let store = self.store.read().await;
        store.is_empty()
    }
    
    /// Removes expired entries from the cache and returns the count of removed entries
    pub async fn cleanup_expired(&self) -> usize {
        let mut store = self.store.write().await;
        let before_len = store.len();
        store.retain(|_, (_, time)| time.elapsed() < self.ttl);
        before_len - store.len()
    }

    /// Saves the cache to disk
    pub async fn save(&self) -> Result<()> {
        let cache_file = self.cache_dir.join("cache.json");
        
        let store = self.store.read().await;
        let serializable: HashMap<String, StoredCacheEntry> = store
            .iter()
            .map(|(k, (v, time))| {
                let created_at = SystemTime::now()
                    .duration_since(SystemTime::UNIX_EPOCH)
                    .unwrap()
                    .as_millis() as u64 - time.elapsed().as_millis() as u64;
                
                (k.clone(), StoredCacheEntry {
                    value: v.clone(),
                    created_at,
                })
            })
            .collect();
            
        let content = serde_json::to_string(&serializable)
            .map_err(|e| ProcessorError::Config(e.to_string()))?;
        
        tokio::fs::write(cache_file, content)
            .await
            .map_err(|e| ProcessorError::IO(e))?;
        
        Ok(())
    }
    
    /// Loads the cache from disk
    pub async fn load(&self) -> Result<()> {
        let cache_file = self.cache_dir.join("cache.json");
        
        if cache_file.exists() {
            let content = tokio::fs::read_to_string(&cache_file)
                .await
                .map_err(|e| ProcessorError::IO(e))?;
            
            let entries: HashMap<String, StoredCacheEntry> = serde_json::from_str(&content)
                .map_err(|e| ProcessorError::Config(e.to_string()))?;
            
            let now = Instant::now();
            let current_time = SystemTime::now()
                .duration_since(SystemTime::UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64;
                
            let mut store = self.store.write().await;
            for (key, entry) in entries {
                // Calculate how old this entry is
                let age = current_time - entry.created_at;
                let age_duration = Duration::from_millis(age);
                
                // Create an Instant that represents when this entry was created
                let creation_instant = if age < now.elapsed().as_millis() as u64 {
                    // Entry is older than our program has been running
                    now - age_duration
                } else {
                    // Handle edge case - use a very old timestamp
                    now - now.elapsed()
                };
                
                store.insert(key, (entry.value, creation_instant));
            }
        }
        
        Ok(())
    }

    /// Invalidates a cache entry by its key
    pub async fn invalidate(&self, key: &str) -> Result<()> {
        let path = self.get_path(key);
        
        if path.exists() {
            // Use tokio::fs instead of std::fs
            tokio::fs::remove_file(&path)
                .await
                .map_err(|e| ProcessorError::IO(e))?;
        }
        
        Ok(())
    }

    /// Stores a value in the cache with the specified key
    pub async fn set(&self, key: &str, value: &str) -> Result<()> {
        let path = self.get_path(key);
        
        if let Some(parent) = path.parent() {
            if !parent.exists() {
                // Use tokio::fs instead of std::fs
                tokio::fs::create_dir_all(parent)
                    .await
                    .map_err(|e| ProcessorError::IO(e))?;
            }
        }
        
        // Use tokio::fs instead of std::fs
        tokio::fs::write(&path, value)
            .await
            .map_err(|e| ProcessorError::IO(e))?;
        
        Ok(())
    }

    /// Stores a value in the cache with the specified key and metadata
    pub async fn set_with_metadata(
        &self, 
        key: &str, 
        value: &str, 
        metadata: &CacheMetadata
    ) -> Result<()> {
        let path = self.get_path(key);
        let meta_path = self.get_metadata_path(key);
        
        let entry = CacheEntry {
            url: key.to_string(),
            path: path.clone(),
            timestamp: std::time::SystemTime::now(),
            ttl: metadata.ttl,
        };
        
        // Use tokio::fs instead of std::fs
        tokio::fs::write(&meta_path, serde_json::to_string(&entry)?)
            .await
            .map_err(|e| ProcessorError::IO(e))?;
        
        self.set(key, value).await
    }

    fn get_metadata_path(&self, key: &str) -> PathBuf {
        let mut path = self.get_path(key);
        path.set_extension("meta");
        path
    }

    fn get_path(&self, key: &str) -> PathBuf {
        // Create a safe filename from the key
        let safe_key = key.replace(['/', ':', '?', '&', '='], "_");
        self.cache_dir.join(safe_key)
    }
}

/// Metadata for cache entries with additional information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CacheMetadata {
    /// How long this entry should remain valid
    pub ttl: Duration,
    /// The source of this cache entry
    pub source: String,
    /// When this entry was created
    pub created_at: chrono::DateTime<chrono::Utc>,
}

impl CacheMetadata {
    /// Creates a new CacheMetadata with the specified TTL and source
    pub fn new(ttl: Duration, source: &str) -> Self {
        Self {
            ttl,
            source: source.to_string(),
            created_at: chrono::Utc::now(),
        }
    }
}

File: python/src/llamapackageservice/cache.py
"""
Caching mechanisms to improve performance.

This module provides both file-based and in-memory caching for
downloaded content and processed results.
"""

from pathlib import Path
from datetime import datetime, timedelta
from typing import TypeVar, Generic, Optional, Dict, Any
import hashlib
import json
import asyncio
from dataclasses import dataclass, field

import aiofiles
import aiofiles.os

from .error import CacheError


@dataclass
class CacheEntry:
    """Represents a single entry in the file cache."""
    
    url: str
    path: Path
    timestamp: datetime
    ttl: timedelta
    
    def is_expired(self) -> bool:
        """Check if this cache entry has expired."""
        return datetime.now() - self.timestamp > self.ttl
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for serialization."""
        return {
            "url": self.url,
            "path": str(self.path),
            "timestamp": self.timestamp.isoformat(),
            "ttl_seconds": self.ttl.total_seconds(),
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "CacheEntry":
        """Create from dictionary."""
        return cls(
            url=data["url"],
            path=Path(data["path"]),
            timestamp=datetime.fromisoformat(data["timestamp"]),
            ttl=timedelta(seconds=data["ttl_seconds"]),
        )


class FileCache:
    """A file-based cache for storing downloaded or generated content."""
    
    def __init__(self, cache_dir: Path, default_ttl: timedelta = timedelta(hours=24)):
        self.cache_dir = Path(cache_dir)
        self.default_ttl = default_ttl
    
    def _get_safe_key(self, key: str) -> str:
        """Create a safe filename from the key."""
        # Replace special characters with underscores
        safe_key = key.replace("/", "_").replace(":", "_").replace("?", "_").replace("&", "_").replace("=", "_")
        # If the key is too long, hash it
        if len(safe_key) > 200:
            safe_key = hashlib.sha256(key.encode()).hexdigest()
        return safe_key
    
    def _get_path(self, key: str) -> Path:
        """Get the cache file path for a key."""
        safe_key = self._get_safe_key(key)
        return self.cache_dir / safe_key
    
    def _get_meta_path(self, key: str) -> Path:
        """Get the metadata file path for a key."""
        return self._get_path(key).with_suffix(".meta")
    
    async def get(self, key: str) -> Optional[CacheEntry]:
        """
        Retrieve a cached entry by its key if it exists and is not expired.
        
        Args:
            key: The cache key to look up
            
        Returns:
            The cached entry if found and not expired, None otherwise
        """
        meta_path = self._get_meta_path(key)
        
        if not await aiofiles.os.path.exists(meta_path):
            return None
        
        try:
            async with aiofiles.open(meta_path, 'r') as f:
                content = await f.read()
            
            entry = CacheEntry.from_dict(json.loads(content))
            
            # Check if entry is expired
            if entry.is_expired():
                # Entry is expired, remove it
                await self.remove(key)
                return None
            
            return entry
        except Exception as e:
            raise CacheError(f"Failed to read cache entry: {e}", source=e)
    
    async def set(
        self,
        key: str,
        value: bytes,
        ttl: Optional[timedelta] = None
    ) -> Path:
        """
        Store a value in the cache with the specified key and TTL.
        
        Args:
            key: The cache key
            value: The content to cache
            ttl: Optional TTL, defaults to default_ttl
            
        Returns:
            Path to the cached file
        """
        path = self._get_path(key)
        meta_path = self._get_meta_path(key)
        
        # Ensure cache directory exists
        if path.parent:
            await aiofiles.os.makedirs(path.parent, exist_ok=True)
        
        try:
            # Write the value
            async with aiofiles.open(path, 'wb') as f:
                await f.write(value)
            
            # Create and save the entry metadata
            entry = CacheEntry(
                url=key,
                path=path,
                timestamp=datetime.now(),
                ttl=ttl or self.default_ttl,
            )
            
            async with aiofiles.open(meta_path, 'w') as f:
                await f.write(json.dumps(entry.to_dict()))
            
            return path
        except Exception as e:
            raise CacheError(f"Failed to write cache entry: {e}", source=e)
    
    async def remove(self, key: str) -> bool:
        """Remove an entry from the cache."""
        path = self._get_path(key)
        meta_path = self._get_meta_path(key)
        removed = False
        
        try:
            if await aiofiles.os.path.exists(path):
                await aiofiles.os.remove(path)
                removed = True
            if await aiofiles.os.path.exists(meta_path):
                await aiofiles.os.remove(meta_path)
                removed = True
        except Exception:
            pass
        
        return removed
    
    async def clear(self) -> int:
        """Clear all entries from the cache. Returns number of entries removed."""
        count = 0
        try:
            if await aiofiles.os.path.exists(self.cache_dir):
                for item in self.cache_dir.iterdir():
                    await aiofiles.os.remove(item)
                    count += 1
        except Exception:
            pass
        return count


T = TypeVar('T')


class Cache(Generic[T]):
    """A generic in-memory cache for any type."""
    
    def __init__(self, ttl: timedelta = timedelta(hours=1)):
        self._store: Dict[str, tuple[T, datetime]] = {}
        self._ttl = ttl
        self._lock = asyncio.Lock()
    
    async def get(self, key: str) -> Optional[T]:
        """
        Retrieve a value from the cache by its key.
        
        Args:
            key: The cache key
            
        Returns:
            The cached value if found and not expired, None otherwise
        """
        async with self._lock:
            if key not in self._store:
                return None
            
            value, timestamp = self._store[key]
            
            # Check if expired
            if datetime.now() - timestamp > self._ttl:
                del self._store[key]
                return None
            
            return value
    
    async def set(self, key: str, value: T) -> None:
        """
        Store a value in the cache with the specified key.
        
        Args:
            key: The cache key
            value: The value to store
        """
        async with self._lock:
            self._store[key] = (value, datetime.now())
    
    async def remove(self, key: str) -> bool:
        """
        Remove an entry from the cache by its key.
        
        Args:
            key: The cache key
            
        Returns:
            True if the entry was removed, False if it didn't exist
        """
        async with self._lock:
            if key in self._store:
                del self._store[key]
                return True
            return False
    
    async def clear(self) -> None:
        """Clear all entries from the cache."""
        async with self._lock:
            self._store.clear()
    
    async def cleanup_expired(self) -> int:
        """
        Remove all expired entries from the cache.
        
        Returns:
            Number of entries removed
        """
        async with self._lock:
            now = datetime.now()
            expired_keys = [
                key for key, (_, timestamp) in self._store.items()
                if now - timestamp > self._ttl
            ]
            for key in expired_keys:
                del self._store[key]
            return len(expired_keys)
    
    def __len__(self) -> int:
        """Return the number of entries in the cache."""
        return len(self._store)


class StringCache(Cache[str]):
    """Specialized cache for string values."""
    pass


class BytesCache(Cache[bytes]):
    """Specialized cache for bytes values."""
    pass
